test/check-machines-migrate

```diff
@@ -0,0 +1,187 @@
+# the Free Software Foundation; either version 2.1 of the License, or
+# (at your option) any later version.
+#
+# Cockpit is distributed in the hope that it will be useful, but
+# WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+# Lesser General Public License for more details.
+#
+# You should have received a copy of the GNU Lesser General Public License
+# along with Cockpit; If not, see <http://www.gnu.org/licenses/>.
+
+import os
+import sys
+
+# import Cockpit's machinery for test VMs and its browser test API
+TEST_DIR = os.path.dirname(__file__)
+sys.path.append(os.path.join(TEST_DIR, "common"))
+sys.path.append(os.path.join(os.path.dirname(TEST_DIR), "bots/machine"))
+
+from machineslib import VirtualMachinesCase  # noqa
+from testlib import wait, test_main  # noqa
+from machinesxmls import POOL_XML  # noqa
+
+
+class TestMachinesMigration(VirtualMachinesCase):
+    provision = {
+        "machine1": {"address": "10.111.113.1/20", "dns": "10.111.112.100"},
+        "machine2": {"address": "10.111.113.2/20", "dns": "10.111.112.100"},
+    }
+
+    def prepare(self, machine1, machine2):
+        machine1.execute("echo '10.111.113.2 machine2' >> /etc/hosts")
+        machine1.execute("echo '10.111.113.1 machine1' >> /etc/hosts")
+        machine2.execute("echo '10.111.113.2 machine2' >> /etc/hosts")
+        machine2.execute("echo '10.111.113.1 machine1' >> /etc/hosts")
+
+        machine2.execute("sed -i 's/127.0.1.1.*/127.0.1.1 machine2 machine2/' /etc/hosts")
+        machine2.execute("hostnamectl set-hostname machine2")
+        machine2.execute("systemctl restart libvirtd")
+
+        # setup key authentication
+        machine1.execute("mkdir -m 0700 -p $HOME/.ssh")
+        machine1.execute("ssh-keygen -t rsa -N '\' -f $HOME/.ssh/id_rsa")
+        machine1.execute("echo 'StrictHostKeyChecking no' > $HOME/.ssh/config")
+        machine1PubKey = machine1.execute("cat $HOME/.ssh/id_rsa.pub")
+        machine2.execute("echo '{0}' >> /root/.ssh/authorized_keys".format(machine1PubKey))
+
+        # Setup the default storage pool on the destination host
+        # This is normally automatically created when creating the first VM but in this case our dest VM was never populated
+        xml = POOL_XML.format(path="/var/lib/libvirt/images")
+        machine2.execute("echo \"{0}\" > /tmp/xml && virsh pool-define /tmp/xml && virsh pool-start images".format(xml))
+
+        # Iptables < 1.8.3 will cause firewall-cmd --reload to fail
+        # https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=914694
+        # Let's disable the firewall on debian-stable and let the test run instead of adding a naughty
+        if machine1.image not in ["debian-stable"]:
+            # Plain migration: The source host VM opens a direct unencrypted TCP connection to the destination host for sending the migration data.
```

Comment:
Which port should we use then? If we let libvirt auto find a port, we still need firewall allow port rule, no? 

I just tried with port 22 and I get:
Error message: operation failed: Failed to connect to remote libvirt URI qemu+tcp://root@machine2:22/system: packet 1397966889 bytes received from server too large, want 33554432

Reply:
So seems listening on tcp is disabled by default: see listen_tcp in the configuration file. I suggest therefore to not test a scenario which is not the default.

Reply:
> Which port should we use then?

Naïvely, none at all. qemu+ssh:// should do *everything* through ssh, for precisely the reason that opening ports (and unencrypted ones even) is painful, insecure, and unfriendly. But that's why I said "sounds like", not "is" -- I don't actually know what libvirt really does. qemu+tcp on port 22 seems odd indeed, as that's ssh's port. If you want to use tcp "plain mode", then yes, you need to pick a port and the test setup is fine.

What confused me is that the test setup *prepares* for tcp://, but the tests always use ssh, and never actually use tcp://. So this sounds like there is some "hidden" communication going on which don't use ssh after all, even when migrating with qemu+ssh:// ?

Reply:
Well, it doesn't use only ssh.
When using qemu+ssh://, the ssh is used to just communicate with libvirt on the other side. Thru ssh it starts the process on the other side, It tells it and prepares that there will be a new VM coming. Everything related to just communication with libvirt.
But the data itself are transferred on lower level thru qemu. And there only tcp or unix can be used. The tcp is used by default.
So basically you need to set --desturi, like qemu+ssh://destination_host.com/system. 
Then optionally you can also set --migrateuri, where you change manually the migration uri to something like: tcp://REMOTE_HOST:PORT . But since REMOTE_HOST can be deducted from desturi, tcp is the option which most users would choose for migration anyway, and it's also wise to let libvirt choose port, it's okay to omit this option in UI.

Reply:
Therefore in tests you can see we are preparing key authentication for qemu+ssh://, so libvirts on source and destination can communicate with each other thru ssh
And then we are opening ports so qemu can migrate the data thru tcp

Reply:
I see, thanks! Would of course be great if libvirt would just set up port forwarding for QEMU, so that this doesn't have to expose an unprotected TCP port. That's not hard to do, and would make this feature so much more useful, but for now that's outside of our control I suppose. But it's another moving part, so the code needs to be prepared for the possibility of ssh working, but qemu/tcp not (... yet another error condition test case).

Reply:
Added a test case